### ISSUE TITLE
Fluid Discretizer on an AE subnet breaks fluid crafting on a main net.

### ISSUE THREAD

**Issue by AbdielKavash** [GTNewHorizons/GT-New-Horizons-Modpack#14808]
### Your GTNH Discord Username

abdiel_kavvash

### Your Pack Version

2.4.1

### Your Server

SP

### Java Version

Java 19

### Type of Server

Single Player

### Your Expectation

My main AE network uses an AE2FC fluid discretizer to facilitate on-demand crafts using fluids.

I would like to build a second network (subnet), which has access to all resources (both items and fluids) from the main net, and is also capable of crafting items on its own automatically, using level maintainers or crafting cards.

I expose the resources on the main network to the subnet using a combination of interface -> storage bus, one pair for items and another pair for fluids. Since the subnet wants to run crafts using fluids as well, it needs its own fluid discretizer.

I expect the subnet to be able to run autonomously, using resources from the main net as needed for crafting.

### The Reality

Since the required setup to replicate this is fairly complex, I am including a video:

https://www.dropbox.com/scl/fi/8j789eer2aln1jbbut4ac/UnlX9C4KrO.mp4?rlkey=xarp0c8ve97cyj9lwytxrikq4&dl=0

* I show an overview of both networks. "Main" network in this setup is the red one on the left, "subnet" is blue on the right. The networks are *not* connected in any way. (Having two discretizers in the same network is *not* the issue here.)
* I show the configuration of the interface -> storage bus combos. All four components are left in their default config.

First I go over the red ("main") network:
* The red network has a fluid crafting pattern for making epoxid sheets.
* (Not shown in the video because I forgot, the storage drive on the red network contains one multi-fluid and one item storage cell.)
* The network contains 4 ingots (576 mB) of molten epoxid and molten rubber, and 5 buckets of water.
* The same amounts are reported in the fluid and item terminals.
* I demonstrate successfully crafting an epoxid sheet in the red network.

Now on to the blue ("subnet") network:
* The blue network has a fluid crafting pattern for making rubber sheets.
* The fluid terminal on the blue network correctly reports 3 ingots of epoxid, 4 ingots of rubber, and 5 buckets of water.
* The item terminal on the blue network however incorrectly reports **double** these amounts.
* Crafting a rubber sheet in the blue network succeeds, and reverts the display of fluids to the correct 3 ingots of each (and 5B of water), in both the item and the fluid terminal.

So far everything works, except for the bugged doubled display amount of fluids.

However, I return to the red network:
* The red network now reports **no** fluids, neither in the fluid nor in the item terminal.
* The fluids are still present in the storage cell, they are just invisible to the terminals.
* This is not just a visual issue: attempting to craft another epoxid sheet **fails**. Even though the crafting simulator reports the molten epoxid as available, the Start button is greyed out and the craft can not be started.
* Inserting a new fluid through the fluid terminal restores the red network to a working state.

### Your Proposal

At least the duplicated display of fluids in the item terminal I can explain: the blue network "sees" the fluid stored in the red network, and the blue discretizer converts this into drops. However, the blue network *also* sees the fluid drops from the red networks, and mistakenly adds these together.

Indeed removing the discretizer from the blue network fixes this first issue; at the cost of not being able to use fluid crafting in the blue network (which defeats the whole purpose of this setup).

The red network fluid storage being made inaccessible is the main problem here, and could be a consequence of the same issue.

I have found one almost-functional workaround: set the fluid storage bus on the blue network to "Insert Only" mode. With this change, the fluid terminal on the blue network does not show any fluids at all. However, the blue item terminal reports the correct amount of fluid drops, and fluid crafting in both networks works correctly. I am assuming that the blue network can still withdraw fluid drops from the red network through the item storage bus and convert them to fluids for crafting using its discretizer. However, since fluids from the red network are no longer accessible to the blue network, the duplicate fluid amount issue does not occur, and the bug breaking crafting in the red network is not triggered.

### Final Checklist

- [X] I have searched this issue tracker and there is nothing similar already. Posting on a closed issue saying the bug still exists will prompt us to investigate and reopen it once we confirm your report.
- [X] I can reproduce this problem consistently by follow the exact steps I described above, or this does not need reproducing, e.g. recipe loophole.
- [ ] I have asked other people and they confirm they also have this problem by follow the exact steps I described above, or this does not need reproducing, e.g. recipe loophole.

**Comment by fluffle** [comment]
I have something similar to this working. I've not watched the video cos I'm on mobile rn, but:

- make sure all storage bus -> interface connections are going one way, I have interfaces on main net, storage buses on level maintainer subnet
- have one fluid and one item storage bus unfiltered, import only. This allows subnet to read main net stuff.
- have 1-N storage buses filtered to the level maintained end products, either fluid or item, export only. This allows subnet to write crafted items to main net.
- export only storage buses need to be higher priority than subnet storage, which is only for storing intermediates (i.e. mixed alloy dust)
- crafting machines like mixers/EBFs etc need to have their me output bus/hatch on the subnet so the crafting CPUs on that subnet can register completed crafts, the outputs will still go to main net via the storage buses.

I'll be on discord tomorrow evening probably if this hasn't been enough for you to get something working.

**Comment by Ydrec** [comment]
Something you probably realize but

Ae doesn't need a descritizer in the network to do fluid crafting, it just needs to see drops. 
This means if your subnet doesn't have descrizer or fluid bus connecting it with the main network it can still use fluids from the main net for crafting provided the descritizer exists in the main.

You can probably add fluid bus connecting subnet > main net so main network descritizer can see subnet stored fluids and create drops for them without needing subnet to have its own.

**Comment by AbdielKavash** [comment]
This works for crafts which take fluids and output items. But for crafts which *produce* fluids, if the network doesn't have a discretizer, it never recognizes fluid produced from the machine entering a dual interface, and it never completes the craft.

**Comment by fluffle** [comment]
Yes, you need a discretizer on both networks. But that should be fine as long as you make sure your interface -> storage bus links are only ever in one direction, to avoid network loops.

**Comment by AbdielKavash** [comment]
Both links are the same, interface on the red network, and storage bus on the blue. Please watch the video in the OP.

**Comment by fluffle** [comment]
I had a look but the audio is incredibly quiet so I struggled to hear your explanations.

It is interesting to observe that only the crafting terminal sees duplication, presumably because it is observing the fluids via both the fluid storage bus directly and via the item storage bus as droplet items, as you said. I hadn't observed this myself because the only item view I have on my maintainer subnet is the ME chest containing a single item storage cell. This suggests another workaround might be to place an inverted fuzzy item filter on the *item* storage bus, preventing it from seeing droplet items? I can't tell from theorycrafting alone whether this would be a better solution than your "insert only" approach.

And actually this might explain something I have observed -- the item droplet representation of certain fluids in my main network disappears and reappears occasionally. I suspect I have not experienced the same level of brokenness as you because there is quite a lot of fluid traffic on my main network and, as you observed, other fluids being reinserted causes an update to whatever is caching the fluid contents as items in the discretizer.
